tv_apps/tv-deck/js/tv_deck.js

```diff
@@ -0,0 +1,538 @@
+   */
+  proto.scanChannels = function td_scanChannels() {
+    var source = this.playingSource.source;
+    this.playingSource.channels = [];
+    this.playingSource.channelIndexHash = {};
+    if (source.isScanning) {
+      return;
+    }
+
+    if (!source.onscanningstatechanged) {
+      source.onscanningstatechanged = function onscanningstatechanged(event) {
+        var state = event.state;
+        switch (state) {
+          case 'completed':
+          /* falls through */
+          case 'stopped':
+            this._onScanningCompleted();
+            break;
+          default:
+            break;
+        }
+      }.bind(this);
+    }
+
+    source.startScanning({
+      isRescanned: true
+    });
+  };
+
+  /**
+   * Retrieve all the currently available TV channels from current TV source.
+   * Sort channels for channel switching.
+   */
+  proto._onScanningCompleted = function td_onScanningCompleted() {
+    this.playingSource.source.getChannels().then(function onsuccess(channels) {
+      if (channels.length === 0) {
+        console.log('Error, no channel found!');
+        return;
+      }
+
+      // Sort channels. Channel number can be XX-XX-XX
+      channels.sort(function(channelA, channelB) {
+        return this._compareChannel(channelA.number, channelB.number);
+      }.bind(this));
+
+      var i;
+      for (i = 0; i < channels.length; i++) {
+        this.playingSource.channelIndexHash[channels[i].number] = i;
+        this.playingSource.channels[i] = {
+          channel: channels[i],
+          programs: {},
+        };
+      }
+
+      if (!this.playingChannelNumber) {
+        this.playingChannelNumber = channels[0].number;
+      }
+
+      if (!this.playingChannel) {
+        this._showErrorState();
+        return;
+      }
+
+      this.setHash();
+    }.bind(this), function onerror(error) {
+      alert(error);
+    });
+  };
+
+  proto.setPlayingSource = function td_setPlayingSource(callback) {
+    if (this.playingTuner.tuner.currentSource === this.playingSource.source) {
+      this.setPlayingChannel(callback);
+      return;
+    }
+
+    this.playingTuner.tuner.setCurrentSource(this.playingSourceType)
+      .then(function() {
+        this.setPlayingChannel(callback);
+      }.bind(this), function() {
+        alert('Source not found');
+      });
+  };
+
+  proto.setPlayingChannel = function td_setPlayingChannel(callback) {
+    this.playingSource.source.setCurrentChannel(this.playingChannelNumber)
+      .then(function() {
+        /* jshint maxlen:false */
+        var newStream = this.playingTuner.tuner.stream;
+        if (this.tvStreamElement.src !== newStream) {
+          this.tvStreamElement.src = newStream;
+          this.tvStreamElement.play();
+        }
+        if (callback) {
+          callback();
+        }
+      }.bind(this), function() {
+        alert('Channel not found');
+      });
+  };
+
+  proto._compareChannel = function td__compareChannel(channelA, channelB) {
```

Comment:
Why do we need to use "String" to compare channel? May we use the tuple(service id, network id, ...) to compare it.